usaspending_api/config/envs/default.py

```diff
@@ -60,57 +67,126 @@ def __new__(cls, *args, **kwargs):
-                postgres_password=values["POSTGRES_PASSWORD"].get_secret_value(),
-                postgres_host=values["POSTGRES_HOST"],
-                postgres_port=values["POSTGRES_PORT"],
-                postgres_db=values["POSTGRES_DB"],
+    DATABASE_URL: str = None  # FACTORY_PROVIDED_VALUE. See below root validator-factory
+    USASPENDING_DB_NAME: str = "data_store_api"
+    USASPENDING_DB_USER: str = ENV_SPECIFIC_OVERRIDE
+    USASPENDING_DB_PASSWORD: SecretStr = ENV_SPECIFIC_OVERRIDE
+    USASPENDING_DB_HOST: str = ENV_SPECIFIC_OVERRIDE
+    USASPENDING_DB_PORT: str = ENV_SPECIFIC_OVERRIDE
+
+    @root_validator
+    def _DATABASE_URL_and_parts_factory(cls, values):
+        """A root validator to backfill DATABASE_URL and USASPENDING_DB_* part config vars and validate that they are all
+        consistent.
+
+        - Serves as a factory function to fill out all places where we track the database URL as both one complete
+        connection string and as individual parts.
+        - ALSO validates that the parts and whole string are consistent. A ``ValueError`` is thrown if found to
+        be inconsistent, which will in turn raise a ``pydantic.ValidationError`` at configuration time.
+        """
+
+        # First determine if DATABASE_URL was provided.
+        is_database_url_provided = values["DATABASE_URL"] and values["DATABASE_URL"] not in CONFIG_VAR_PLACEHOLDERS
+
+        # If DATABASE_URL was provided
+        # - it should take precedence
+        # - its values will be used to backfill any missing POSTGRES DSN parts stored as separate config vars
+        if is_database_url_provided:
+            url_parts, username, password = parse_pg_uri(values["DATABASE_URL"])
+
+            # Backfill only POSTGRES DSN CONFIG vars that are missing their value
+            if "USASPENDING_DB_HOST" not in values or values["USASPENDING_DB_HOST"] is None:
+                values = eval_default_factory_from_root_validator(
+                    cls, values, "USASPENDING_DB_HOST", lambda: url_parts.hostname
+                )
+            if "USASPENDING_DB_PORT" not in values or values["USASPENDING_DB_PORT"] is None:
+                values = eval_default_factory_from_root_validator(
+                    cls, values, "USASPENDING_DB_PORT", lambda: str(url_parts.port)
+                )
+            if "USASPENDING_DB_NAME" not in values or values["USASPENDING_DB_NAME"] is None:
+                values = eval_default_factory_from_root_validator(
+                    cls, values, "USASPENDING_DB_NAME", lambda: url_parts.path.lstrip("/")
+                )
+            if "USASPENDING_DB_USER" not in values or values["USASPENDING_DB_USER"] is None:
+                values = eval_default_factory_from_root_validator(cls, values, "USASPENDING_DB_USER", lambda: username)
+            if "USASPENDING_DB_PASSWORD" not in values or values["USASPENDING_DB_PASSWORD"] is None:
+                values = eval_default_factory_from_root_validator(
```

Comment:
Is it possible to replace these with python dict `.get()` coalesce equivalents? Ex:
`"USASPENDING_DB_HOST" not in values or values["USASPENDING_DB_HOST"] is None`
=>
`values.get("USASPENDING_DB_HOST", None) is None`

Reply:
Good idea! changing ...